src/box/box.cc

```diff
@@ -5381,10 +5421,20 @@ bootstrap(bool *is_bootstrap_leader)
 
 	while (true) {
 		box_restart_replication();
-		master = replicaset_find_join_master();
-		bootstrap_leader_uuid = master == NULL ? INSTANCE_UUID :
-							 master->uuid;
-		box_broadcast_ballot();
+		/*
+		 * If the bootstrap strategy is 'supervised' and
+		 * it is known that the given instance is the
+		 * bootstrap leader, just bootstrap the database.
+		 *
+		 * Otherwise, find who is the bootstrap leader.
+		 */
+		if (bootstrap_strategy != BOOTSTRAP_STRATEGY_SUPERVISED ||
```

Comment:
> If I have two instances, where I called box.ctl.make_bootstrap_leader() before box.cfg, they won't be able to join with each other, since each of them supposes itself as bootstrap leader, they will be bootstrapped independently and then fail to join with ER_REPLICASET_UUID_MISMATCH. The question here, are we ok with this behavior? @sergepetrenko @Totktonada

This matches current `make_bootstrap_leader()` behavior. The only difference is that now you may call `box.ctl.make_bootstrap_leader()` before issuing `box.cfg`. But otherwise the user can shoot off his legs even now: nothing prevents him from issuing `make_bootstrap_leader()` on both nodes while they are still trying to connect to each other.

Reply:
> they will be bootstrapped independently and then fail to join with `ER_REPLICASET_UUID_MISMATCH`. The question here, are we ok with this behavior?

I'm not really OK, but OK as a temporary state.

I would like to have some safety checks, but since we have no them now (in this bootstrap strategy), I think it is OK to go step-by-step.

> @Totktonada Did you consider introducing a function in config module, which will save, whether we're supposed to be a bootstrap leader and then call `make_bootstrap_leader()` right after launching box.cfg in a separate fiber? 

Sure. I don't like how much logic is coupled with a certain point in the code, where we call `box.cfg()`. Letting box remember the choice allows to write the same logic simpler.

But it is not the only reason to propose this patch. I also can't bootstrap a single instance replicaset without an iproto port in the supervised strategy and need to switch between config/supervised depending on some conditions.

To sum up: I don't like how complicated/fragile config's code would be without this change. I'm afraid that I'll miss more scenarios in attempting to work around current supervised strategy constraints.

Reply:
> This matches current make_bootstrap_leader() behavior. The only difference is that now you may call box.ctl.make_bootstrap_leader() before issuing box.cfg. But otherwise the user can shoot off his legs even now: nothing prevents him from issuing make_bootstrap_leader() on both nodes while they are still trying to connect to each other.

Now they must be called almost simultaneously for that to happen. When it's done before box.cfg it so much easier to shoot off the legs. Moreover, user don't have any way to remove that `am_i_bootstrap_leader` flag, he cannot check, whether the instance will become bootstrap_leader before box.cfg.

We may consider adding internal function for now, which will set this flag. But If everyone is ok with how it works now, then I don't mind.

@Totktonada, can we consider logging, when `box.ctl.make_bootstrap_leader` is called? It won't be very fun to debug on prod, when some failover calls `make_bootstrap_leader` on several instances.

Reply:
> @Totktonada, can we consider logging, when `box.ctl.make_bootstrap_leader` is called? It won't be very fun to debug on prod, when some failover calls `make_bootstrap_leader` on several instances.

I like this idea. Will add.

Reply:
Added a test for `ER_REPLICASET_UUID_MISMATCH`. Logs are to be added (later today).

Reply:
Added the logs in the second commit.

I didn't bother about other bootstrap strategies, only adjusted code that seems relevant to the supervised strategy. I hope it is enough in context of this PR.